### PR TITLE
[ROCm] Add prequantized FP8 attention for Qwen3 Next

### PR DESCRIPTION
**Pull request by nholmber**
## Purpose

Add a reusable prequantized-QKV attention contract and wire Qwen3 Next on ROCm to AITER fused Q/K RMSNorm, RoPE, gate extraction, FP8 quantization, and FP8 FMHA.

The implementation preserves BF16 model-visible Q/K/V while forwarding FP8 Q/K/V and per-sequence/per-KV-head descales to attention. Cached context can remain FP8 during direct gathering; BF16 KV caches are supported by quantizing only the gathered active context.

## Changes

- Add the generic `PrequantizedQKV` attention-layer contract.
- Forward optional prequantized Q/K/V through the attention API.
- Add ROCm AITER handling for prequantized Q/K/V and descales.
- Integrate AITER `fused_qk_norm_rope_gate_fp8_quant` with Qwen3 Next.
- Keep the preparation path CUDA-graph safe.
- Gather cached FP8 context directly without a BF16 staging round trip.
- Support BF16 KV cache by quantizing only the gathered active context.
- Fall back to the existing BF16 path when `max_num_seqs` exceeds AITER capacity.
- Register `VLLM_ROCM_FP8_DIRECT_CONTEXT_GATHER` in `vllm.envs`.
- Add focused unit and ROCm integration coverage, including mixed batches and multiple KV heads.

## Dependency

Requires ROCm/aiter#5002.

## Validation

Validated after rebasing onto current `upstream/main` using the updated AITER branch on MI355X/gfx950:

```text
focused vLLM tests: 46 passed
AITER focused GPU tests: 18 passed
ruff check: passed
ruff format --check: passed
mypy/pre-commit hooks: passed
git diff --check: passed
```

Earlier end-to-end validation:

```text
long-context and packed coherence: passed
GSM8K: 97.57% flexible-extract
GPQA-Diamond: 92.42%
FP8 ASM FMHA dispatch: confirmed

BF16 KV + gather-time FP8 + FP8 ASM versus BF16 attention:
  wall time: -17.14%
  attention time: -76.30%
```

## AI assistance disclosure

This change was developed with OpenAI Codex assistance. The human submitter must review every changed line and confirm they understand and can defend the implementation before marking the PR ready.

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/runner/moe_runner.py [hunk 1]
```diff
@@ -887,7 +888,14 @@ def _forward_impl(
...
                 self._maybe_fuse_gate_weights()
-                router_logits = F.linear(hidden_states, self._combined_gate_weight)
+                # Route through the platform GEMM dispatcher rather than
+                # F.linear: on ROCm the un-fused branch below reaches the
+                # skinny-GEMM kernels via ReplicatedLinear, and calling
+                # F.linear here would silently drop the fused gate onto
+                # generic rocBLAS instead.
+                router_logits = dispatch_unquantized_gemm()(
+                    self, hidden_states, self._combined_gate_weight, None
+                )
```
Comment (jhartika-amd) [inline vllm/model_executor/layers/fused_moe/runner/moe_runner.py:898]: Could this change be done actually as a separate PR? Seems it's not really related to attention which the topic of this PR. It could also land before the required AITER PR gets merged.

File: vllm/model_executor/layers/qwen3_next_fp8_qkv.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+)
+from vllm.platforms import current_platform
+from vllm.utils.torch_utils import (
+    LayerNameType,
+    _encode_layer_name,
+    _resolve_layer_name,
+    direct_register_custom_op,
+)
+
+MAX_PREQUANTIZED_SEQUENCES = 256
```
Comment (simondanielsson) [inline vllm/model_executor/layers/qwen3_next_fp8_qkv.py:19]: **Question**: What happens if max_num_seqs>256? Is it instead possible to pass along say `min(256, max_num_seqs)` to the aiter call and still have it be graph capture safe while supporting larger max_num_seqs

File: vllm/v1/attention/backends/rocm_aiter_fa.py [hunk 3]
```diff
@@ -35,6 +37,21 @@
 
 _PARTITION_SIZE_ROCM = 256
 _CP_TOKENS_PER_ITER_ROCM = 32 * 1024
+
+
+def _fp8_context_gather_enabled() -> bool:
+    return os.environ.get(
+        "VLLM_ROCM_FP8_DIRECT_CONTEXT_GATHER",
```
Comment (simondanielsson) [inline vllm/v1/attention/backends/rocm_aiter_fa.py:44]: **Suggestion**: Should we add this to envs.py?

File: vllm/model_executor/models/qwen3_next.py [hunk 4]
```diff
@@ -695,6 +768,8 @@ def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
 
 
 class QwenNextMixtureOfExperts(MixtureOfExperts):
+    num_local_physical_experts: int
```
Comment (simondanielsson) [inline vllm/model_executor/models/qwen3_next.py:771]: **Nit**: is this strictly needed?

**Review by simondanielsson — COMMENTED** [review]
Nice work!

File: vllm/model_executor/layers/qwen3_next_fp8_qkv.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+            gate,
+            *_allocate_prequantized_outputs(
+                q_gate,
+                key,
+                value,
+                num_query_heads,
+                num_kv_heads,
+                head_dim,
+            ),
+        )
```
Comment (damien-lejeune) [inline vllm/model_executor/layers/qwen3_next_fp8_qkv.py]: Here `*_allocate_prequantized_outputs` returns empty (i.e. uninitialized) tensors, probably to serve as placeholder due to the inflexible interface. It's okay, but these unitialized tensors are packed into a `PrequantizedQKV`, then `self.attn(...)` is called (Vllm's Aiter integration) with it. On the Vllm's Aiter integration side, it uses `q_descale=None` (in the decode branch of `AiterFlashAttentionImpl`'s `_forward()` method at`vllm/v1/attention/backends/rocm_aiter_fa.py`). So now it works okay. But it only works because the integration is implemented to not consume the `q_descale` in `prequantized_qkv: PrequantizedQKV`. There is probably a structural reason why the decode path in the integration is not consuming it though. Nonetheless, this is quite bug-prone to have this sort of unenforced contract IMHO. Is this something you think we could improve? Or at least having a comment acknowledging this contract?

File: tests/model_executor/layers/test_rocm_aiter_prequantized_qkv.py [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+    ],
+)
+def test_aiter_backend_gathers_context_for_fp8_attention(
+    monkeypatch,
+    kv_cache_dtype,
+    cache_dtype,
+    expected_quantize,
+):
+    impl = object.__new__(rocm_aiter_fa.AiterFlashAttentionImpl)
+    impl.num_kv_heads = 1
```
Comment (damien-lejeune) [inline tests/model_executor/layers/test_rocm_aiter_prequantized_qkv.py:194]: Could we also have a test for when `num_kv_heads > 1`? Here we just have a mock kernel, but it would still be nice to test for this case to increase coverage for things outside of the kernel (e.g. the wiring the integration work this PR adds).

File: tests/model_executor/layers/test_rocm_aiter_prequantized_qkv.py [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+        query_start_loc=torch.tensor([0, 2], dtype=torch.int32),
+        max_seq_len=2,
+        seq_lens=torch.tensor([2], dtype=torch.int32),
+        slot_mapping=torch.tensor([0, 1], dtype=torch.int64),
+        block_table=torch.zeros((1, 1), dtype=torch.int32),
+        causal=True,
+        num_decodes=0,
+        num_decode_tokens=0,
+        num_prefills=1,
+        num_extends=0,
```
Comment (damien-lejeune) [inline tests/model_executor/layers/test_rocm_aiter_prequantized_qkv.py:126]: """
It looks like we are only testing the case for `(num_decodes=0, num_extends=0, num_prefill=1)` here. Could we add a test (or parametrization) covering `(num_decodes>0, num_extends>0, num_prefill>0)` to check the descale offset arithmetic?
"""

File: vllm/model_executor/layers/fused_moe/runner/moe_runner.py (hunk 1 above)
Comment (jhartika-amd) [reply vllm/model_executor/layers/fused_moe/runner/moe_runner.py]: @mjkvaak-amd has this as an own PR: https://github.com/vllm-project/vllm/pull/54185

Comment (nholmber) [reply vllm/model_executor/layers/fused_moe/runner/moe_runner.py]: Removed from this PR. The shared-expert dispatcher change is tracked separately in #54185.

File: vllm/model_executor/layers/fused_qk_norm_rope_gate_fp8_quant.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+)
+from vllm.platforms import current_platform
+from vllm.utils.torch_utils import (
+    LayerNameType,
+    _encode_layer_name,
+    _resolve_layer_name,
+    direct_register_custom_op,
+)
+
+MAX_PREQUANTIZED_SEQUENCES = 256
```
Comment (nholmber) [reply vllm/model_executor/layers/fused_qk_norm_rope_gate_fp8_quant.py:19]: Addressed by gating the integration on the configured scheduler limit. If `max_num_seqs > 256`, this optimization is disabled and the existing BF16 path is used; clamping the value would be unsafe for a runtime batch above 256.

File: vllm/v1/attention/backends/rocm_aiter_fa.py (hunk 3 above)
Comment (nholmber) [reply vllm/v1/attention/backends/rocm_aiter_fa.py]: Done. `VLLM_ROCM_FP8_DIRECT_CONTEXT_GATHER` is now registered in `vllm/envs.py`, and the backend reads it through `vllm.envs`.

File: vllm/model_executor/models/qwen3_next.py (hunk 4 above)
Comment (nholmber) [reply vllm/model_executor/models/qwen3_next.py]: No; removed the redundant class annotation.

File: vllm/model_executor/layers/qwen3_next_fp8_qkv.py (hunk 5 above)
Comment (nholmber) [reply vllm/model_executor/layers/qwen3_next_fp8_qkv.py]: Addressed explicitly. The contract now documents that decode-only entries may be undefined and must not be consumed. I also added a regression test using poisoned prequantized buffers that verifies pure decode uses the BF16 query and KV cache.

File: tests/model_executor/layers/test_rocm_aiter_prequantized_qkv.py (hunk 7 above)
Comment (nholmber) [reply tests/model_executor/layers/test_rocm_aiter_prequantized_qkv.py:165]: Added mixed decode/extend/prefill offset coverage. The backend now uses one helper to slice token data and the matching per-sequence descales, and the test checks the extend and prefill offsets independently.

File: tests/model_executor/layers/test_rocm_aiter_prequantized_qkv.py (hunk 6 above)
Comment (nholmber) [reply tests/model_executor/layers/test_rocm_aiter_prequantized_qkv.py]: Added coverage for both one and two KV heads in the direct context-gather tests, including per-head scale forwarding.
